### DISCUSSION TITLE
How to change the default location for nu config files?

### DISCUSSION THREAD

**Discussion by ShaiAvr** [nushell/nushell#13472] [Q&A] [answered]
I just installed `nu` on my Windows desktop and the first thing that happens when I launch it is that it detects there is no existing config file and asks me to create one on `C:\Users\username\AppData\Roaming\nushell\env.nu`. I don't want to create those files in the `AppData` folder. I'd rather have it in a more accessible location like `~/.config` or something like that. I couldn't find how I can configure the default location `nu` looks for config files. How can I change it to generate the configuration files into a custom location?

**Comment by fdncred** [comment, marked as answer]
Look for XDG_CONFIG_HOME here https://www.nushell.sh/book/configuration.html#nushell-configuration-with-env-nu-and-config-nu

**Reply by ShaiAvr** [reply]
It worked. Thank you. How would I set this environment variable on my MacBook (or any Unix/Linux device)?
I usually define environment variables in `.zshrc` or a similar config file, but that is only relevant if I launch `nu` from `zsh`. If I am not mistaken, setting `nu` to my default shell will mean that `.zshrc` won't be processed anymore, so where do I put this environment variable?
The link you provided specifically states that it must be set before starting Nushell so it can't be inside `env.nu`.

**Reply by ghost** [reply]
[systemd-wide env-vars exist](https://help.ubuntu.com/community/EnvironmentVariables#System-wide_environment_variables)

or via your desktop ([kde](https://userbase.kde.org/Session_Environment_Variables), [window-managers](https://wiki.debian.org/Xinitrc), etc)

or (what i use) via [home-manager](https://nix-community.github.io/home-manager/)

**Reply by fdncred** [reply]
or in the terminal configuration

**Reply by ShaiAvr** [reply]
@fdncred I couldn't find any configuration of the environment variables in my terminal application (I am using iTerm2). I don't think the terminal is supposed to be responsible for the environment variables, or perhaps I misunderstood your answer.

**Reply by ShaiAvr** [reply]
@Jan9103 Your instructions apply to Ubuntu and other Linux distros. As I stated, I am using a MacBook with MacOS, so your methods don't work for me.

**Reply by ghost** [reply]
apple seems to have a user-space file at `~/.launchd.conf`, which dosnt work(?).  
and a system wide at `/etc/launchd.conf`, which could break a lot in a multi-user environment (even if the other users are only services)

<https://stackoverflow.com/questions/135688>

**Reply by ShaiAvr** [reply]
Now I am more confused than before 😅.

**Reply by fdncred** [reply]
> @fdncred I couldn't find any configuration of the environment variables in my terminal application (I am using iTerm2). I don't think the terminal is supposed to be responsible for the environment variables, or perhaps I misunderstood your answer.

It's pretty common these days. Many terminals do it. Here's a few that I've used. I'm using Ghostty right now and it doesn't support this functionality either yet.
https://alacritty.org/config-alacritty.html
https://wezfurlong.org/wezterm/config/lua/config/term.html
https://raphamorim.io/rio/docs/next/configuration-file/#environment-variables


Some people work around this functionality by using bash or zsh and as the last line in their config start nushell.

**Reply by ghost** [reply]
> Now I am more confused than before 😅.

ill try to explain the whole thing:

environmental variables are general purpose configuration values.  
they can be used to tell programs which c-compiler to use or in this case where the user-specific program-configuration-directory (`XDG_CONFIG_HOME`) is located at.  
environmental variables get passed to child-processes. meaning if you start firefox from a terminal firefox will use the env-vars set in your terminal.  
if you want to set the `XDG_CONFIG_HOME` for nushell you just have to have it set (and not overwritter later on) in any parent-process of it.  
the programs, which are basically guaranteed on a desktop (not server) to be parents of your shell are: the init-system (launchd for apple, systemd/openrc/.. on linux), the window-manager/desktop-environment (kde, gnome, i3, etc), and your terminal (iTerm2, wezterm, etc).  
one thing you have to watch out for is that we are trying to set the **USER SPECIFIC** program-configuration-directory. if we were to set this for all users it might break things for other users on the device.  
other users don't have to be humans, but can exist in the form of services as well. on my desktop for example i have a user which remaps my keyboard-layout, a user running a music-program, a user running a anti-virus, a root user, etc.  

macos is in general very closed and doesnt give you a lot of easy control over non-gui parts of your system.  
in this case (as far as i was able to find) they only offer the option to set system-wide (meaning for all user) env-vars (since their user-specific option is broken).  
therefore the best option for macos is probably to change your terminal configuration.  

if iTerm2 really has no config for this its probably best to switch to a proper terminal and configure it there.  
i used to use wezterm (cross-platform), but switched since it broke frequently on wayland.  
i heard a lot of good things about alacritty.

**Reply by ShaiAvr** [reply]
@Jan9103 Thank you for the detailed explanation. I think I will switch to wezterm (I heard many good things about it and since I work on Mac and Windows, I like it being cross-platform).

Although I must admit it's quite strange for me to define environment variables in the terminal emulator itself. Maybe I am just used to Windows where environment variables are defined on the system level (there is a built-in GUI to define system and user variables), but it makes more sense to me that variables are defined for the system (or user) so the parent process shouldn't need to deal with it.
